src/gap_to_julia.jl

```diff
@@ -22,233 +22,205 @@ of converted objects and should never be given by the user.
-    elseif Globals.IsList( obj )
-        return Array{UInt8,1}(map(i->gap_to_julia(UInt8,obj[i]),1:length(obj)))
-    else
-        throw(ArgumentError("<obj> is not a list"))
     end
+    Globals.IsList(obj) && return [UInt8(obj[i]) for i in 1:length(obj)]
+    throw(ConversionError(obj, Vector{UInt8}))
+end
+
+function guess_type(obj::GapObj) # TODO
+    Globals.IsList(obj) && return Vector{guess_type(first(obj))}
+    Globals.IsRecord(obj) && return Dict{Symbol, Any}
+    # (...) list of singleton types here
+    return Any
 end
+guess_type(::T) where T = T
 
 ## Arrays
-function gap_to_julia( ::Type{Array{Obj,1}}, obj :: GapObj , recursion_dict = IdDict() )
-    if ! Globals.IsList( obj )
-        throw(ArgumentError("<obj> is not a list"))
-    end
-    if haskey(recursion_dict,obj)
-        return recursion_dict[obj]
-    end
-    len_list = length(obj)
-    new_array = Array{Any,1}( undef, len_list)
-    recursion_dict[obj] = new_array
-    for i in 1:len_list
-        new_array[i] = ElmList(obj, i)  # returns 'nothing' for holes in the list
-    end
-    return new_array
+function Base.Vector(obj::GapObj, recursion_dict=IdDict())
+    Globals.IsList(obj) || throw(ConversionError(obj, Vector))
+    T = Union{(guess_type(obj[i]) for i in 1:length(obj))...}
+    return Vector{T}(obj, recursion_dict)
 end
 
-function gap_to_julia( ::Type{Array{T,1}}, obj :: GapObj, recursion_dict = IdDict() ) where T
-    if ! Globals.IsList( obj )
-        throw(ArgumentError("<obj> is not a list"))
-    end
-    if haskey(recursion_dict,obj)
-        return recursion_dict[obj]
-    end
-    len_list = length(obj)
-    new_array = Array{T,1}( undef, len_list)
-    recursion_dict[obj] = new_array
-    for i in 1:len_list
-        current_obj = ElmList(obj, i)
-        new_array[i] = get!(recursion_dict, current_obj) do
-            gap_to_julia(T, current_obj, recursion_dict)
+function Base.Matrix(obj::GapObj, recursion_dict=IdDict())
+    Globals.IsList(obj) || throw(ConversionError(obj, Vector))
+    f = first(obj)
+    Globals.IsList(f)  || throw(ConversionError(obj, Vector))
+    T = guess_type(f)
+    return Matrix{T}(obj, recursion_dict)
+end
+
+# shallow layer for recursion:
+_gap_to_julia(::Type, obj, recursion_dict) = obj
+_gap_to_julia(::Type{T}, obj::GapObj, recursion_dict) where T =
+    T(obj, recursion_dict)
+_gap_to_julia(::Type{Any}, obj::GapObj, recursion_dict) = gap_to_julia(obj)
+
+function Base.Vector{T}(obj::GapObj, recursion_dict=IdDict()) where T
+    Globals.IsList(obj) || throw(ConversionError(obj, Vector{T}))
+
+    if !haskey(recursion_dict, obj)
+        new_array = Vector{T}(undef, length(obj))
+        for i in 1:length(obj)
+            current_obj = ElmList(obj, i)
+            # should return 'missing' for holes in the list
+            new_array[i] = get!(recursion_dict, current_obj,
+                _gap_to_julia(T, current_obj, recursion_dict))
         end
+        recursion_dict[obj] = new_array
```

Comment:
Why shall the assignment in `recursion_dict` be moved behind the loop?
If `obj` is self-referential then the changed variant runs into an error, whereas the old code had no problem.
Here is an example.
```
l = GAP.Globals.EvalString( GAP.julia_to_gap( "[ 1, ~, 3 ]" ) );
GAP.gap_to_julia( l )
```

Reply:
yep, I didn't even consider self-referring lists
btw.
```julia
julia> s = GAP.julia_to_gap( "[ 1, ~, 3 ]" )
GAP: "[ 1, ~, 3 ]"

julia> GAP.Globals.EvalString(s)
Error, Panic: tried to print a path of unsupported type 'component object'
Error showing value of type Main.ForeignGAP.MPtr:
ERROR: Error thrown by GAP
(...)
```
crashes

Reply:
As discussed on issue #324, the issue is in printing. But one can just suppress that:
```julia
julia> x = @gap [1,~,3] ;

julia> GAP.gap_to_julia(x)
3-element Array{Any,1}:
 1
  Any[1, Any[#= circular reference @-1 =#], 3]
 3

```